File: solutions/swift/leap/1/Leap.swift
//
//  Leap.swift
//  Leap
//
//  Created by Barrett Clark on 7/2/14.
//  Copyright (c) 2014 Barrett Clark. All rights reserved.
//

import Foundation

class Year {
  var calendarYear: Int = 0

  init(calendarYear: Int) {
    self.calendarYear = calendarYear
  }
  
  var isLeapYear: Bool {
    get {
      return isCentury() ? isExceptionalCentury() : isVanillaLeapYear()
    }
  }
  
  func isVanillaLeapYear() -> Bool { return self.calendarYear % 4 == 0 }
  func isCentury() -> Bool { return self.calendarYear % 100 == 0 }
  func isExceptionalCentury() -> Bool { return self.calendarYear % 400 == 0 }
}

File: solutions/typescript/leap/1/leap.ts
export function isLeap(year: number): boolean {
  return isCentury(year) ? isExceptionalCentury(year) : isVanillaLeap(year);
}

function isYearDivisibleBy(year: number, n: number): boolean {
  return year % n === 0;
}

function isCentury(year: number): boolean {
  return isYearDivisibleBy(year, 100);
}

function isVanillaLeap(year: number): boolean {
  return isYearDivisibleBy(year, 4);
}

function isExceptionalCentury(year: number): boolean {
  return isYearDivisibleBy(year, 400);
}
